aten/src/ATen/core/dispatch/DispatchKeyExtractor.h

```diff
@@ -105,6 +105,14 @@ struct CAFFE2_API DispatchKeyExtractor final {
     return DispatchKeyExtractor(0);
   }
 
+  void registerSchema(const FunctionSchema& schema) {
```

Comment:
1. FunctionSchema still seems right, it lets Sebastian, e.g., decide to store a bitmap instead of arity, without having to plumb a constructor change as well in the patch.
2. I agree. But I don't want to add a guard inside the boxed calling hotpath. Maybe we can arrange for operator lookup to NOT return the operator if it doesn't have a schema.
3. No, `def_count` can be greater than 1, because legacy registrations repeatedly register schema.

Reply:
> 1. FunctionSchema still seems right, it lets Sebastian, e.g., decide to store a bitmap instead of arity, without having to plumb a constructor change as well in the patch.

Avoiding the constructor change works for me, but I'm not sure what you mean about deciding to store a bitmap vs arity - can you point me to where that's done? All I'm seeing is `DispatchTable` forwarding the schema to `DispatchKeyExtractor`, which just sets `num_args` off it then throws it away...

> 2. I agree. But I don't want to add a guard inside the boxed calling hotpath. Maybe we can arrange for operator lookup to NOT return the operator if it doesn't have a schema.

Agree, the guard should be "only ops in valid state should be addressable" rather than "make sure the op is in valid state when calling". 

That would also lend itself to a concurrency-friendly idiom where you load all the impls first, then commit the schema.

> 3. No, `def_count` can be greater than 1, because legacy registrations repeatedly register schema.

Ah I misread the code, despite your good comments :) are we committed to ongoing support for that behavior? (This is the LIFO-behavior-in-notebooks use case, right?)

Reply:
+1 on (2), i.e. only having valid ops being findable

Reply:
> Avoiding the constructor change works for me, but I'm not sure what you mean about deciding to store a bitmap vs arity - can you point me to where that's done? All I'm seeing is DispatchTable forwarding the schema to DispatchKeyExtractor, which just sets num_args off it then throws it away...

It's a hypothetical change, not in master, but one that has been done in PRs before. Here's a quote: https://github.com/pytorch/pytorch/pull/33313/files#diff-9ce4d20732d7b4c47e6a95c566f83ff6R101

> are we committed to ongoing support for that behavior? (This is the LIFO-behavior-in-notebooks use case, right?)

Not quite. People who were using the old RegisterOperators API always induce a "def" situation. My BC plan here is: (1) migrate all uses we have direct control over to the new API, (2) make it invalid to mix new-style with old-style registration, but preserve multiple defs for old api, (3) delete the old API entirely, at which point def can only happen once. I don't know how long it will take to get to (3) though.

The notebook stuff is for overriding impls. The plan on record here is to add a priority system and make notebook loading use priority to control ordering. I don't plan on doing this in the near future.

Reply:
Deprecation plan at https://github.com/pytorch/pytorch/issues/35040

Reply:
Jupyter plan at https://github.com/pytorch/pytorch/issues/35041

Reply:
> It's a hypothetical change, not in master, but one that has been done in PRs before. Here's a quote: [#33313 (files)](https://github.com/pytorch/pytorch/pull/33313/files#diff-9ce4d20732d7b4c47e6a95c566f83ff6R101)

Ah of course. I know about this change, just didn't make the connection that it would be happening here. Thanks!

Thanks also for the legacy API and Jupyter disambiguation.

Reply:
This is fixed now! findSchema only returns ops with schema. findOperatorByName returns everything.